plugins/async/src/main/java/org/apache/struts2/async/DefaultAsyncManager.java

```diff
@@ -0,0 +1,149 @@
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.struts2.async;
+
+import com.opensymphony.xwork2.AsyncManager;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.struts2.ServletActionContext;
+
+import javax.servlet.AsyncContext;
+import javax.servlet.AsyncEvent;
+import javax.servlet.AsyncListener;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Implements {@link AsyncManager} to add support for invoke async actions via Servlet 3's API.
+ *
+ * @since 2.6
+ */
+public class DefaultAsyncManager implements AsyncManager, AsyncListener {
+    private static final Logger LOG = LogManager.getLogger(DefaultAsyncManager.class);
+    private static final AtomicInteger threadCount = new AtomicInteger(0);
+
+    private AsyncContext asyncContext;
+    private boolean asyncActionStarted;
+    private Boolean asyncCompleted;
+    private Object asyncActionResult;
+
+    @Override
+    public void invokeAsyncAction(final Callable asyncAction) {
+        if (asyncActionStarted) {
+            return;
+        }
+
+        Long timeout = null;
+        Executor executor = null;
+        if (asyncAction instanceof AsyncAction) {
+            AsyncAction customAsyncAction = (AsyncAction) asyncAction;
+            timeout = customAsyncAction.getTimeout();
+            executor = customAsyncAction.getExecutor();
+        }
+
+        HttpServletRequest req = ServletActionContext.getRequest();
+        asyncActionResult = null;
+        asyncCompleted = false;
+
+        if (asyncContext == null || !req.isAsyncStarted()) {
+            asyncContext = req.startAsync(req, ServletActionContext.getResponse());
+            asyncContext.addListener(this);
+            if (timeout != null) {
+                asyncContext.setTimeout(timeout);
+            }
+        }
+        asyncActionStarted = true;
+        LOG.debug("Async processing started for " + asyncContext);
+
+        final Runnable task = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    setAsyncActionResultAndDispatch(asyncAction.call());
+                } catch (Throwable e) {
+                    setAsyncActionResultAndDispatch(e);
+                }
+            }
+        };
+        if (executor != null) {
+            executor.execute(task);
+        } else {
+            final Thread thread = new Thread(new Runnable() {
```

Comment:
Maybe we should just throw an `Exception` here to force App developers to take care of their `Executor`. WDYT?

Reply:
No, providing an executor is optional. This if statement is not a fall back on user mistake but means an executor is optional :)

Reply:
And I wonder if it would be better to have it not optional

Reply:
I saw such in spring framework during my elementary studies and I think they have made a good decision to add this option. I think this is an additional option not a default needed one.

Reply:
Then we come back to my first comment: it is important to provide a sane default to avoid this else branch

Reply:
No, I meant default executor is a thread created and started by Struts. If user did not like, then can provide an executor as wish :)

Reply:
> I meant default executor is a thread created and started by Struts

I want to avoid this. The whole async topic is about performance. Createing a single thread is expensive and slow. That's why thread pools and executors exist. Having slow default behavior in a performance critical situation is a bad thing.

Reply:
> If apps enter this else branch things are likely slow. So we should provide defaults to avoid this. I cannot find where executor is created. Could we initialize it in AsyncAction?

I'm sorry, I , now, accidentally saw above in my mails. I don't know why it's not present here!

Anyway, now, I understand your worry. Please don't worry as I checked other frameworks like Spring and saw they also have such behavior as default and works. Please see [WebAsyncManager.java:177](https://github.com/spring-projects/spring-framework/blob/6c3a64578c10dd4e6fed933864f1721cf35203ae/spring-web/src/main/java/org/springframework/web/context/request/async/WebAsyncManager.java#L77) and then [SimpleAsyncTaskExecutor.java:236](https://github.com/spring-projects/spring-framework/blob/b6cae219d824f4990d77cac61477316ec678a09b/spring-core/src/main/java/org/springframework/core/task/SimpleAsyncTaskExecutor.java#L236).

For now, the importance is to provide scaleability by just simply running action outside of the container. For performance, if I add thread pool, then I will have to enable user to configure it (number of threads, scheduling mechanism, etc). somehow a lot of work :( I think, as an introduction, we can keep it simple as is. When needed, user can create and configure an executor finely as wish. If really they complain, then we have to add these. let what happens :)